README.md
# parcel-plugin-pwa-manifest
[![Build Status](https://travis-ci.com/101arrowz/parcel-plugin-pwa-manifest.svg?branch=master)](https://travis-ci.com/101arrowz/parcel-plugin-pwa-manifest) [![Coverage Status](https://coveralls.io/repos/github/101arrowz/parcel-plugin-pwa-manifest/badge.svg?branch=master)](https://coveralls.io/github/101arrowz/parcel-plugin-pwa-manifest?branch=master) [![tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier) [![built with TypeScipt](https://badgen.net/badge/types/TypeScript)](https://www.typescriptlang.org)

A Parcel plugin that generates a Web App Manifest, creates all the icons you need, and more!

## Usage
In `package.json`:
```json
{
  "name": "my-awesome-pwa",
  "description": "An awesome PWA to do awesome things",
  "pwaManifest": {
    "name": "My Awesome PWA",
    "shortName": "My PWA",
    "startURL": "./offline",
    "theme": "#add8e6",
    "generateIconOptions": {
      "baseIcon": "./public/my-awesome-icon.svg",
      "sizes": [192, 384, 512],
      "genFavicons": true
    }
  }
}
```

This will create a `manifest.webmanifest` similar to the following:
```json
{
  "name": "My Awesome PWA",
  "short_name": "My PWA",
  "description": "An awesome PWA to do awesome things",
  "start_url": "./offline",
  "theme_color": "#add8e6",
  "icons": [
    {
      "src": "./my-awesome-icon-192x192.webp",
      "sizes": "192x192",
      "type": "image/webp"
    },
    {
      "src": "./my-awesome-icon-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "./my-awesome-icon-384x384.webp",
      "sizes": "384x384",
      "type": "image/webp"
    },
    {
      "src": "./my-awesome-icon-384x384.png",
      "sizes": "384x384",
      "type": "image/png"
    },
    {
      "src": "./my-awesome-icon-512x512.webp",
      "sizes": "512x512",
      "type": "image/webp"
    },
    {
      "src": "./my-awesome-icon-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ]
}
```
In `index.html`, a link to the manifest, an Apple Touch Icon, a Microsoft Tile configuration, and two favicons will be inserted at the top of the `<head>`.

## What?
This package is a plugin for the [Parcel bundler](https://parceljs.org) that creates a web manifest with reasonable defaults and inserts a link into the HTML. More importantly, it handles all icon/favicon generation for you when you provide a base icon, which lets you effortlessly support multiple screen sizes and ensure best practices.

## Why?
Parcel fully supports web manifests, but creating them can be annoying and can involve you having to jump through hoops to do simple things such as having an icon set that just works across all devices.

In the case of icon generation, you could manually do it with something like [Real Favicon Generator](https://realfavicongenerator.net), but if you ever change your main icon, you have to run through the entire process again. That's no fun.

Integrating manifest generation into the build pipeline makes life easier. Only minimal configuration is *required*, but if you want you can still customize to your heart's content.

## Are there any limitations?
Yes. As of now, the plugin only supports applications that have a single HTML file, `index.html`, at their root paths. That should be almost all of them, but if you need the manifest served at multiple HTML files, you're out of luck. (If anyone needs this, you can make an issue and provide your use case.)

Safari pinned tab support would also require an autotracer to get an SVG output, which is too large an addition for a small plugin like this. If you absolutely need pinned tab support, you have to manually insert it into the original HTML file.

## Documentation
All configuration is done in `package.json` under the `pwaManifest` key.

Anything that usually goes in a `manifest.json` file besides `icons` can go into `pwaManifest`. All parameter names have aliases (e.g. `start_url` can also be `startURL`, `startUrl`, or `start-url`). If you see any inconsistencies in the documentation, it's probably fine; you can use multiple names for the same value.

### Changes from standard manifest
The `name` and `description` will default to their counterparts in the outer layer of `package.json`.

The `theme_color` will default to white and will change the default behavior of some parts of the icon generation, such as the background color of the Microsoft Tile.

Instead of manually setting an `icons` parameter containing a set of icons, you should use `genIconOpts`, which will contain the options for icon generation. The parameters for `genIconOpts` are as follows:
#### `baseIcon`
The path to the icon to generate all other icons from. Path is relative to the placement of `package.json`.
- For best results, use a high-resolution (at least 512x512) PNG or an SVG.
#### `sizes`
An array of pixel values for the sizes to generate. Defaults to `[96, 152, 192, 384, 512]`.
- All PWAs need at least 192px and 512px icons, so those will be added in regardless of whether they are provided in the `sizes` parameter.
#### `formats`
An object whose keys are the desired output formats (in lowercase) and whose values are the configurations to use with the [`sharp`](https://sharp.pixelplumbing.com/en/stable/api-output/#png) package when generating icons of that type. By default, generates WebP and PNG images with somewhat high compression.
- If you prefer the user getting one format over another, put that format first in the object.
- All PWAs need at least PNG, so that's inserted first if you provide your own config. If you don't want it first, put your own `png` key-value pair in your config.
#### `appleTouchIconBG`
The background color for the Apple Touch Icon (to fill transparent regions). Defaults to the theme color.
- Useful because by default, Apple uses black for transparent regions, which doesn't look good with most icons.
#### `appleTouchIconPadding`
The number of pixels to pad the Apple Touch Icon with on all sides. Defaults to 12.
- Used to account for Apple's courner-rounding.
#### `genFavicons`
Whether or not to generate 16x16 and 32x32 favicons and insert links in the HTML. Defaults to `false`.
#### `msTileColor`
The background color for Microsoft Tiles. Defaults to the theme color.
#### `resizeMethod`
The method to use for resizing non-square images. Can be one of `'cover'` (default), `'contain'`, or `'fill'`.

---

All parameters that exist in the [MDN documentation for the Web App Manifest](https://developer.mozilla.org/en-US/docs/Web/Manifest) are aliased, type-checked, and insterted into the manifest whenever provided in `pwaManifest`. There are a few reasonable defaults, like `'.'` for `start_url`. 

If you need to have a parameter not included in that list, put an array of parameter names to keep in the final manifest under the `include` key.

## Known Issues
- JPEG output has a black background by default. There's nothing I can do about this other than add a new option for background color because JPEG does not support transparency.
  - I suggest that you don't output JPEG at all, I only offer it for those who may need it. 

## License
MIT

CHANGELOG.md
## 2.5.11-2.5.13
- Boosted rank
## 2.5.9-2.5.10
- Improved README
## 2.5.8
- Fixed tests (should not have been a version change)
- Finished adding all tests
## 2.5.7
- Overhauled testing mechanism
- Added coverage report
## 2.5.6
- Minor improvements
## 2.5.5
- Added badge, license
## 2.5.4
- Add tests
  - Hopefully no more broken versions
## 2.5.3
- Respect public URL provided (broken before)
## 2.5.2
- Switched to ambient type declarations (should not have updated version)
## 2.5.1
- Fixed 2.5.0 bugs
## 2.4.0-2.5.0
- Added support for multiple entry ports
- Deprecated because broken
## 2.3.0
- Minor bugfixes (should have been 2.2.2)
## 2.2.1
- Used `unknown` instead of implicity `any` type
## 2.2.0
- Migrated to TypeScript
- Minor bugfixes
## 2.1.1-2.1.5
- Fixed buggy, failing code
## 2.1.0
- Added Wide MS Tile support
- Added `resizeMethod` option
- Fixed theme color issues
- Updated Parcel minimum version
## 2.0.3
- Respect `contentHash` Parcel option
- Update README with better format
## 2.0.2
- Fixed README format
### 2.0.1
- Created README with documentation
## 2.0.0
- Now an actual PWA manifest generator, rather than [what it was before](https://github.com/101arrowz/parcel-plugin-precache-manifest)

index.test.js
const Bundler = require('parcel-bundler');
const { tmpdir } = require('os');
const attachManifestGenerator = require('./test-build');
const { join } = require('path');
const { readdirSync, readFileSync } = require('fs');
const outDir = join(tmpdir(), `_parcel-plugin-pwa-manifest/${Date.now().toString(36)}`);
const defaultResult = ['apple-touch-icon.105de2c5.png','icon-152x152.c0cccc53.webp','icon-152x152.c5e4d8cb.png','icon-192x192.3103561d.webp','icon-192x192.cab29cd7.png','icon-384x384.2b65be8b.png','icon-384x384.a54ab4a2.webp','icon-512x512.5c88cabb.webp','icon-512x512.b24e0d51.png','icon-96x96.00714ae7.webp','icon-96x96.b73322f2.png','mstile-150x150.df894f37.png','mstile-310x150.23817a3d.png','mstile-310x310.5f619e76.png','mstile-70x70.3bcf1bff.png'];
const testConfigs = [
  {
    msg: 'Default generation works correctly',
    config: {
      generateIconOptions: {
        baseIcon: './sample/icon.svg'
      }
    },
    result: defaultResult
  },
  {
    msg: 'Favicon generation works correctly',
    config: {
      generateIconOptions: {
        baseIcon: './sample/icon.svg',
        genFavicons: true
      }
    },
    result: [...defaultResult, 'favicon-16x16.726e3e17.png', 'favicon-32x32.6326f01e.png']
  }
];
const bundler = new Bundler(join(__dirname, 'sample', 'index.html'), {
  publicUrl: './',
  outDir,
  cache: false,
  watch: false,
  hmr: false,
  contentHash: true,
  logLevel: 1
});
jest.setTimeout(30000);
attachManifestGenerator(bundler);
const manifestGen = bundler.listeners('bundled')[0];
bundler.off('bundled', manifestGen);
const testConfig = ({ config, result, msg, manifest, logOutput }) =>
  test.concurrent(msg || 'icons are correctly generated', () => {
    return new Promise(resolve => {
      bundler.on('bundled', async bundle => {
        bundle.entryAsset.getPackage = async () => ({
          name: 'tester',
          description: 'test',
          pwaManifest: config,
          pkgdir: __dirname
        });
        await manifestGen(bundle);
        const generatedFiles = readdirSync(outDir);
        if (logOutput) console.log(msg+': '+JSON.stringify(generatedFiles));
        expect(generatedFiles).toEqual(expect.arrayContaining([...new Set(result.concat('browserconfig.xml', 'index.html', 'manifest.webmanifest'))]));
        if (manifest) {
          const newManifest = JSON.parse(readFileSync(join(outDir, 'manifest.webmanifest')).toString());
          expect(newManifest).toMatchObject(manifest);
        }
        resolve();
      });
    })
  });
for (const conf of testConfigs) {
  testConfig(conf);
}
bundler.bundle();

__tests__/main.test.ts
import { tmpdir } from 'os';
import { join } from 'path';
import attachManifestGenerator from '../index';
import {
  readdirSync,
  readFileSync,
  writeFileSync,
  mkdirSync,
  existsSync
} from 'fs';
import { EventEmitter } from 'events';
import logger from '@parcel/logger';
import testConfigs, {
  Config,
  ResultConfig,
  BaseConfig,
  ErrorConfig
} from './testConfigs';
jest.mock('@parcel/logger');
const mockedLogger = logger as jest.Mocked<typeof logger>;
// Not quite a mock since it has completely different behavior to original, but should work for all purposes
class Bundler extends EventEmitter {
  private fakeBundle: unknown;
  private manifestGen: (bundle: unknown) => Promise<void>;
  public options: CorrectedParcelOptions & { outDir: string };

  constructor({
    pkgdir = join(__dirname, '..', 'sample'),
    ...opts
  }: Omit<PackageJSON, 'pkgdir'> & {
    pkgdir?: string;
  }) {
    super();
    this.fakeBundle = {
      entryAsset: {
        getPackage: (): Promise<PackageJSON> =>
          Promise.resolve({
            pkgdir,
            ...opts
          })
      }
    };
    const timeString = Date.now().toString(36);
    let outDir = join(tmpdir(), '_parcel-plugin-pwa-manifest', timeString);
    try {
      mkdirSync(outDir);
    } catch (e) {
      // Travis CI
      const tmpDir = join(__dirname, 'tmp');
      if (!existsSync(tmpDir)) mkdirSync(tmpDir);
      outDir = join(tmpDir, timeString);
      mkdirSync(outDir);
    }
    writeFileSync(join(outDir, 'index.html'), '<head></head>');
    this.options = {
      outDir,
      target: 'browser',
      publicURL: './',
      contentHash: true
    };
    attachManifestGenerator((this as unknown) as FullBundler);
    this.manifestGen = this.listeners('bundled')[0] as Bundler['manifestGen'];
    this.off('bundled', this.manifestGen);
  }
  async genIcons(): Promise<void> {
    return this.manifestGen(this.fakeBundle);
  }
  getFileList(): string[] {
    return readdirSync(this.options.outDir);
  }
  getManifest(): unknown {
    return JSON.parse(
      readFileSync(join(this.options.outDir, 'manifest.webmanifest')).toString()
    );
  }
  getBrowserConfig(): string {
    return readFileSync(join(this.options.outDir, 'browserconfig.xml'))
      .toString()
      .slice(74, -39);
  }
  getHTML(): string {
    return readFileSync(join(this.options.outDir, 'index.html'))
      .toString()
      .slice(6, -7);
  }
}
jest.setTimeout(30000);
const DEFAULT_REQUIRED_FILES = [
  'index.html',
  'manifest.webmanifest',
  'browserconfig.xml'
];
const testConfig = ({ config, msg, ...props }: Config): void =>
  test(msg || 'icons are correctly generated', async done => {
    const bundler = new Bundler({
      name: 'tester',
      description: 'test',
      pwaManifest: config
    });
    await bundler.genIcons();
    const errorCall = mockedLogger.error.mock.calls[0];
    if (errorCall && errorCall.length) {
      const error = errorCall[0].slice(26);
      const throws = (props as Omit<ErrorConfig, keyof BaseConfig>).throws;
      if (throws && new RegExp(throws).test(error)) return done();
      throw new Error(error);
    }
    const { result, manifest, browserconfig, html, logOutput } = props as Omit<
      ResultConfig,
      keyof BaseConfig
    >;
    const generatedFiles = bundler.getFileList();
    if (logOutput)
      console.log(
        msg +
          '\nFiles: [' +
          generatedFiles
            .filter(str => !DEFAULT_REQUIRED_FILES.includes(str))
            .map(str => "'" + str.replace(/'/, "\\'") + "'")
            .join(', ') +
          ']\nOutput at: ' +
          bundler.options.outDir
      );
    expect(generatedFiles).toEqual(
      expect.arrayContaining([
        ...new Set(result.concat(...DEFAULT_REQUIRED_FILES))
      ])
    );
    if (manifest) {
      expect(bundler.getManifest()).toMatchObject(manifest as object);
    }
    if (browserconfig) {
      expect(bundler.getBrowserConfig()).toMatch(new RegExp(browserconfig));
    }
    if (html) {
      expect(bundler.getHTML()).toMatch(new RegExp(html));
    }
    done();
  });
for (const conf of testConfigs) {
  testConfig(conf);
}


__tests__/index.test.js
const Bundler = require('parcel-bundler');
const { tmpdir } = require('os');
const attachManifestGenerator = require('..');
const { join } = require('path');
const { readdirSync } = require('fs');
const outDir = join(tmpdir(), `_parcel-plugin-pwa-manifest/${Date.now().toString(36)}`);
const testConfigs = require('./testConfigs.json');
const bundler = new Bundler(join(__dirname, '..', 'sample', 'index.html'), {
  publicUrl: './',
  outDir,
  cache: false,
  watch: false,
  hmr: false,
  logLevel: 1
});
jest.setTimeout(30000);
attachManifestGenerator(bundler);
const manifestGen = bundler.listeners('bundled')[0];
bundler.off('bundled', manifestGen);
const testConfig = (config, resultFiles, msg) =>
  test(msg || 'icons are correctly generated', done => {
    // Simulate package.json
    bundler.on('bundled', async bundle => {
      bundle.entryAsset.getPackage = async () => ({
        name: 'tester',
        description: 'test',
        pwaManifest: config,
        pkgdir: join(__dirname, '..')
      });
      await manifestGen(bundle);
      const generatedFiles = readdirSync(outDir);
      expect(generatedFiles).toEqual(expect.arrayContaining(resultFiles));
      done();
    });
  });
for (const conf of testConfigs) {
  testConfig(conf.config, conf.result, conf.msg);
}
bundler.bundle();

index.ts
import { writeFileSync, existsSync, readFileSync } from 'fs';
import { resolve, basename, sep } from 'path';
import { createHash } from 'crypto';
import { ParcelAsset } from 'parcel-bundler';
import logger from '@parcel/logger';
import sharp, { ResizeOptions } from 'sharp';

// TODO: Add Safari Pinned Tab SVG - could prove to be challenging
export = (bundler: FullBundler): void => {
  let { outDir, publicURL, contentHash, target } = bundler.options;
  // istanbul ignore next
  if (target !== 'browser' || process.env.DISABLE_PWA_MANIFEST) {
    bundler.on('buildEnd', () => logger.warn('Manifest creation disabled'));
    return;
  }
  const hashedFilename = (filename: string, buf: Buffer): string => {
    const i = filename.lastIndexOf('.');
    const base = filename.slice(0, i);
    const ext = filename.slice(i + 1);
    return (
      base +
      '.' +
      createHash('md5')
        .update(
          contentHash
            ? ext + buf.toString()
            : '_parcel-plugin-pwa-manifest' + sep + filename
        ) // Need unique filepath :/
        .digest('hex')
        .slice(-8) +
      '.' +
      ext
    ); // Similar to (but not the same as) Parcel itself
  };
  // istanbul ignore next
  if (!publicURL) publicURL = '/';
  else if (!publicURL.endsWith('/')) publicURL += '/';
  const getPkg = (entryAsset: ParcelAsset): Promise<PackageJSON> =>
    typeof entryAsset.getPackage === 'function'
      ? entryAsset.getPackage()
      : Promise.resolve(entryAsset.package);
  // istanbul ignore next
  const err = (msg: string): void => {
    logger.clear();
    logger.error('Manifest creation failed! ' + msg);
  };
  bundler.on('bundled', async bundle => {
    const pkg = await getPkg(
      bundle.entryAsset || bundle.childBundles.values().next().value.entryAsset
    );
    // istanbul ignore next
    if (!outDir) outDir = resolve(pkg.pkgdir, 'dist');

    const opts = pkg.pwaManifest || pkg['pwa-manifest'];
    // istanbul ignore next
    if (typeof opts !== 'object') {
      if (typeof opts === 'undefined')
        return err('No PWA Manifest options found in package.json.');
      return err(
        'The PWA Manifest parameter in package.json must be an object containing the desired parameters.'
      );
    }

    const insertInto = resolve(outDir, 'index.html'); // TODO: support insertion into multiple HTML files
    // istanbul ignore next
    if (!existsSync(insertInto))
      return err(
        'No index.html found at the root of the build. This package does not yet support this scenario.'
      );

    const name = opts.name || pkg.name;
    // istanbul ignore next
    if (typeof name !== 'string') {
      if (typeof name === 'undefined')
        return err('No name was found in the options.');
      return err('The name provided in the options must be a string.');
    }
    const shortName =
      opts.shortName || opts['short-name'] || opts['short_name'] || name;
    // istanbul ignore next
    if (typeof shortName !== 'string')
      return err('The short name provided in the options must be a string.');

    const desc = opts.desc || opts.description || pkg.description || '';
    // istanbul ignore next
    if (typeof desc !== 'string')
      return err('The description provided in the options must be a string.');

    const startURL =
      opts.startUrl ||
      opts.startURL ||
      opts['start-url'] ||
      opts['start_url'] ||
      publicURL;
    // istanbul ignore next
    if (typeof startURL !== 'string')
      return err('The start URL provided in the options must be a string.');

    const scope = opts.scope || publicURL;
    // istanbul ignore next
    if (typeof scope !== 'string')
      return err('The scope provided in the options must be a string.');

    const theme =
      opts.theme_color ||
      opts.theme ||
      opts.themeColor ||
      opts['theme-color'] ||
      'white';
    // istanbul ignore next
    if (typeof theme !== 'string')
      return err(
        'The theme color provided in the options must be a string representing a valid CSS color.'
      );

    logger.progress(`Generating icons for ${name}...`);

    const icons: Array<IconEntry> = [];
    const genIconOpts =
      opts.genIcon ||
      opts['gen-icon'] ||
      opts.genIconOpts ||
      opts['gen-icon-opts'] ||
      opts.genIconOptions ||
      opts['gen-icon-options'] ||
      opts.generateIcons ||
      opts['generate-icons'] ||
      opts.generateIconOptions ||
      opts['generate-icon-options'] ||
      opts.icons;
    // istanbul ignore next
    if (typeof genIconOpts !== 'object') {
      if (typeof genIconOpts === 'undefined')
        return err(
          'No icon generation options found in the PWA manifest options.'
        );
      return err(
        'The icon generation options in the PWA manifest options must be an object containing the desired parameters.'
      );
    }

    const msTileColor: unknown =
      genIconOpts.msTileColor ||
      genIconOpts['ms-tile-color'] ||
      genIconOpts.microsoftTileColor ||
      genIconOpts['microsoft-tile-color'] ||
      theme;
    // istanbul ignore next
    if (typeof msTileColor !== 'string')
      return err(
        'The Microsoft tile color provided in the options must be a string representing the theme color for the application.'
      );
    let browserConfig = `<TileColor>${msTileColor}</TileColor>`;
    let htmlOut = `<meta name="msapplication-config" content="${publicURL}browserconfig.xml"><meta name="theme-color" content="${theme}">`;

    const baseIconPath: unknown =
      genIconOpts.baseIcon ||
      genIconOpts['base-icon'] ||
      genIconOpts.fromIcon ||
      genIconOpts['from-icon'];
    // istanbul ignore next
    if (typeof baseIconPath !== 'string') {
      if (typeof baseIconPath === 'undefined')
        return err('No base icon was found in the icon generation options.');
      return err(
        'The base icon parameter in the icon generation options must be a string that contains the path to the icon.'
      );
    }

    const baseIconName = basename(
      baseIconPath,
      baseIconPath.slice(baseIconPath.lastIndexOf('.'))
    );
    const baseIconFullPath = resolve(pkg.pkgdir, baseIconPath);
    // istanbul ignore next
    if (!existsSync(baseIconFullPath))
      return err(
        'No icon was found at the base icon path ' + baseIconPath + '.'
      );

    let sizes = [96, 152, 192, 384, 512]; // Common sizes
    // istanbul ignore next
    if (
      genIconOpts.sizes instanceof Array &&
      genIconOpts.sizes.every((v: unknown) => typeof v === 'number')
    )
      sizes = [...new Set(genIconOpts.sizes.concat(192, 512))] as number[];
    // Needed in all PWAs
    else if (typeof genIconOpts.sizes !== 'undefined')
      return err(
        'The sizes parameter in the icon generation options must be an array of numeric pixel values for sizes of the images.'
      );

    let formats: FormatOptions = {
      webp: {
        quality: 60,
        reductionEffort: 6
      },
      png: {
        compressionLevel: 9
      }
    };
    // istanbul ignore next
    if (
      genIconOpts.formats instanceof Object &&
      Object.keys(genIconOpts.formats).every(v =>
        ['png', 'jpeg', 'webp', 'tiff'].includes(v)
      )
    )
      formats = {
        png: formats.png,
        ...genIconOpts.formats
      };
    // PNG needed in all PWAs
    else if (typeof genIconOpts.formats !== 'undefined')
      return err(
        'The formats parameter in the icon generation options must be an object with each key being a supported file type (png, webp, jpeg, or tiff) for the output images, and each value being the options to pass to sharp.'
      );

    const resizeMethod: unknown =
      genIconOpts.resizeMethod ||
      genIconOpts['resize-method'] ||
      genIconOpts.resize ||
      'cover';
    // istanbul ignore next
    if (!['cover', 'contain', 'fill'].includes(resizeMethod as string))
      return err(
        "The resize method parameter in the icon generation options must be one of 'cover', 'contain', or 'fill'."
      );
    const resizeOptions: ResizeOptions = {
      fit: resizeMethod as 'cover' | 'contain' | 'fill',
      background: 'rgba(0, 0, 0, 0)'
    };
    const baseIcon = sharp(baseIconFullPath).ensureAlpha();
    for (const size of sizes) {
      const icon = baseIcon.clone().resize(size, size, resizeOptions);
      const saveSize = size + 'x' + size;
      for (let format of Object.keys(formats) as Array<keyof FormatOptions>) {
        // Ugly but TS is not very smart
        format = format;
        let buf;
        try {
          buf = await icon
            .clone()
            [format](formats[format])
            .toBuffer();
        } catch (e) {
          // istanbul ignore next
          return err(
            'An unknown error ocurred during the icon creation process: ' + e
          );
        }
        const filename = hashedFilename(
          baseIconName + '-' + saveSize + '.' + format,
          buf
        );
        writeFileSync(resolve(outDir, filename), buf);

        icons.push({
          src: publicURL + filename,
          sizes: saveSize,
          type: 'image/' + format
        });
      }
    }
    logger.progress('Generating Apple Touch Icon...');
    const appleTouchIconBG: unknown =
      genIconOpts.appleTouchIconBG ||
      genIconOpts.appleTouchIconBg ||
      genIconOpts['apple-touch-icon-bg'] ||
      genIconOpts.appleTouchIconBackground ||
      genIconOpts['apple-touch-icon-background'] ||
      genIconOpts.atib ||
      theme;
    // istanbul ignore next
    if (typeof appleTouchIconBG !== 'string')
      return err(
        'The Apple Touch Icon background color parameter must be a string representing a valid CSS color.'
      );

    const appleTouchIconPadding: unknown =
      genIconOpts.appleTouchIconPadding ||
      genIconOpts['apple-touch-icon-padding'] ||
      genIconOpts.atip ||
      12;
    // istanbul ignore next
    if (typeof appleTouchIconPadding !== 'number')
      return err(
        'The Apple Touch Icon padding parameter must be a number of pixels to pad the image with on each side.'
      );

    let appleTouchIconBuf: Buffer;
    const atiSize = 180 - 2 * appleTouchIconPadding;
    try {
      const appleTouchIconTransparent = await baseIcon
        .clone()
        .resize(atiSize, atiSize, resizeOptions)
        .extend({
          top: appleTouchIconPadding,
          bottom: appleTouchIconPadding,
          left: appleTouchIconPadding,
          right: appleTouchIconPadding,
          background: 'rgba(0, 0, 0, 0)'
        })
        .toBuffer();
      appleTouchIconBuf = await sharp(appleTouchIconTransparent)
        .flatten({ background: appleTouchIconBG })
        .png(formats.png || {})
        .toBuffer();
    } catch (e) {
      // istanbul ignore next
      return err(
        'An unknown error ocurred during the Apple Touch Icon creation process: ' +
          e
      );
    }
    const atiname = hashedFilename('apple-touch-icon.png', appleTouchIconBuf);
    writeFileSync(resolve(outDir, atiname), appleTouchIconBuf);
    htmlOut += `<link rel="apple-touch-icon" sizes="180x180" href="${publicURL +
      atiname}">`;
    const genFavicons: unknown =
      genIconOpts.genFavicons ||
      genIconOpts['gen-favicons'] ||
      genIconOpts.generateFavicons ||
      genIconOpts['generate-favicons'];
    // istanbul ignore next
    if (!['boolean', 'undefined'].includes(typeof genFavicons))
      return err(
        'The favicon generation option in the icon generation options must be a boolean.'
      );
    if (genFavicons) {
      logger.progress('Generating favicons...');
      for (const size of [32, 16]) {
        let favicon: Buffer;
        try {
          favicon = await baseIcon
            .clone()
            .resize(size, size, resizeOptions)
            .png(formats.png || {})
            .toBuffer();
        } catch (e) {
          // istanbul ignore next
          return err(
            'An unknown error ocurred during the favicon creation process: ' + e
          );
        }
        const sizes = size + 'x' + size;
        const filename = hashedFilename('favicon-' + sizes + '.png', favicon);
        writeFileSync(resolve(outDir, filename), favicon);
        htmlOut += `<link rel="icon" sizes="${sizes}" href="${publicURL +
          filename}">`;
      }
    }
    logger.progress('Generating Microsoft Tile Icons...');
    for (const size of [70, 150, 310]) {
      let msTile: Buffer;
      try {
        msTile = await baseIcon
          .clone()
          .resize(size, size, resizeOptions)
          .png(formats.png || {})
          .toBuffer();
      } catch (e) {
        // istanbul ignore next
        return err(
          'An unknown error ocurred during the Microsoft Tile Icon creation process: ' +
            e
        );
      }
      const sizes = size + 'x' + size;
      const filename = hashedFilename('mstile-' + sizes + '.png', msTile);
      writeFileSync(resolve(outDir, filename), msTile);
      browserConfig += `<square${sizes}logo src="${publicURL + filename}"/>`;
    }
    let rectMsTile: Buffer;
    try {
      rectMsTile = await baseIcon
        .clone()
        .resize(310, 150, resizeOptions)
        .png(formats.png || {})
        .toBuffer();
    } catch (e) {
      // istanbul ignore next
      return err(
        'An unknown error ocurred during the Microsoft Tile Icon creation process: ' +
          e
      );
    }
    const rectMsTileFilename = hashedFilename('mstile-310x150.png', rectMsTile);
    writeFileSync(resolve(outDir, rectMsTileFilename), rectMsTile);
    browserConfig += `<wide310x150logo src="${publicURL +
      rectMsTileFilename}"/>`;

    logger.progress('Generating Microsoft config...');
    writeFileSync(
      resolve(outDir, 'browserconfig.xml'),
      `<?xml version="1.0" encoding="utf-8"?><browserconfig><msapplication><tile>${browserConfig}</tile></msapplication></browserconfig>`
    );

    logger.progress('Generating manifest...');
    // No custom modifications for the rest of the common parameters, so we just do type checking
    const extraParams: PWAManifestOptions = {};
    const extraTypes: [string[], string | Verifier, unknown?][] = [
      [
        [
          'background_color',
          'backgroundColor',
          'background-color',
          'bgColor',
          'bg-color',
          'bg'
        ],
        'string',
        theme
      ],
      [
        ['categories'],
        v => v instanceof Array && v.every(el => typeof el === 'string')
      ],
      [
        ['dir', 'direction', 'textDirection', 'text-direction'],
        v => ['rtl', 'ltr', 'auto'].includes(v as string)
      ],
      [
        ['display', 'displayMode', 'display-mode'],
        v =>
          ['standalone', 'minimal-ui', 'fullscreen', 'browser'].includes(
            v as string
          ),
        'standalone'
      ],
      [
        [
          'iarc_rating_id',
          'iarc',
          'iarcId',
          'iarcID',
          'iarc-id',
          'iarcRatingId',
          'iarcRatingID',
          'iarc-rating-id',
          'iarcRating',
          'iarc-rating'
        ],
        'string'
      ],
      [['lang', 'language'], 'string'],
      [
        ['orientation', 'rotated', 'screenOrientation', 'screen-orientation'],
        v =>
          [
            'any',
            'natural',
            'landscape',
            'landscape-primary',
            'landscape-secondary',
            'portrait',
            'portrait-primary',
            'portrait-secondary'
          ].includes(v as string)
      ],
      [
        [
          'prefer_related_applications',
          'preferRelated',
          'prefer-related',
          'preferRelatedApplications',
          'prefer-related-applications'
        ],
        'boolean'
      ],
      [
        [
          'related_applications',
          'related',
          'relatedApplications',
          'related-applications'
        ],
        v =>
          v instanceof Array && v.every(el => typeof el === 'object' && el.url)
      ],
      [
        ['screenshots', 'screenShots', 'screen-shots'],
        v =>
          v instanceof Array && v.every(el => typeof el === 'object' && el.src)
      ],
      [
        ['serviceworker', 'sw', 'serviceWorker', 'service-worker'],
        v => typeof v === 'object' && !!v && v.hasOwnProperty('src')
      ]
    ];
    for (const type of extraTypes) {
      let val;
      for (const paramName of type[0]) {
        val = opts[paramName];
        if (val) break;
      }
      if (typeof val === 'undefined') {
        if (type[2])
          // Default
          extraParams[type[0][0]] = type[2];
        continue;
      }
      let checker: Verifier;
      if (typeof type[1] === 'string') checker = v => typeof v === type[1];
      else checker = type[1];
      // istanbul ignore next
      if (!checker(val))
        return err(
          'Parameter "' +
            type[0][0] +
            '" provided in the options is invalid. Please check the official MDN documentation on the Web App Manifest.'
        );
      extraParams[type[0][0]] = val;
    }

    // When this config inevitably becomes outdated, use the include parameter to include any new parameters relevant to the Web App Manifest.
    const include: unknown =
      opts.include || opts.includeParams || opts['include-params'];
    // istanbul ignore next
    if (include instanceof Array && include.every(v => typeof v === 'string')) {
      for (const param of include) {
        extraParams[param] = opts[param];
      }
    } else if (typeof include !== 'undefined')
      return err(
        'The include parameter in the options must be an array of extra parameter names to include in the final manifest.'
      );
    const manifest = {
      name,
      short_name: shortName,
      start_url: startURL,
      scope,
      ...(desc && { description: desc }),
      icons,
      theme_color: theme,
      ...extraParams
    };
    writeFileSync(
      resolve(outDir, 'manifest.webmanifest'),
      JSON.stringify(manifest)
    );
    htmlOut += `<link rel="manifest" href="${publicURL}manifest.webmanifest">`;
    logger.progress('Inserting links into HTML...');
    let html = readFileSync(insertInto)
      .toString()
      .replace(
        /<link rel="(manifest|icon|apple-touch-icon)"(.*?)>|<meta name="(msapplication(.*?)|theme-color)"(.*?)>/g,
        ''
      );
    const insertBefore = html.search(/(?<=<head>)|<\/head>/); // Prefer at the start, but end is acceptable too.
    // istanbul ignore next
    if (insertBefore === -1)
      return err(
        'Could not find head tag in HTML file and therefore cannot insert necessary HTML.'
      );
    html = html.slice(0, insertBefore) + htmlOut + html.slice(insertBefore);
    writeFileSync(insertInto, html);
    logger.success('Manifest creation successful.');
  });
};
